File: indexer/runner.py
from __future__ import annotations

import hashlib
import json
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from indexer.config import Config
from indexer.github_client import GitHubClient, GitHubRequestError
from indexer.merge import apply_overrides, load_overrides
from indexer.parser import build_entry, extract_colorschemes
from indexer.state import StateStore


def _safe_repo(repo: str) -> str:
    return repo.strip().removesuffix(".git").strip("/")


def discover_repositories(client: GitHubClient, config: Config) -> dict[str, str]:
    discovered: dict[str, str] = {}

    for topic in config.topics:
        page = 1
        while True:
            items, has_next = client.search_repositories(topic, page, config.per_page)
            if not items:
                break

            for item in items:
                repo = _safe_repo(item.repo)
                if repo and repo not in discovered:
                    discovered[repo] = item.updated_at

            page += 1
            if config.max_pages_per_topic and page > config.max_pages_per_topic:
                break
            if not has_next:
                break

    for repo in config.include_repos:
        normalized = _safe_repo(repo)
        if normalized and normalized not in discovered:
            discovered[normalized] = ""

    return discovered


def _sort_entries(
    entries: list[dict[str, Any]], config: Config
) -> list[dict[str, Any]]:
    reverse = config.sort_order == "desc"

    if config.sort_by == "name":
        key = lambda item: (item.get("name") or "").lower()
    elif config.sort_by == "updated_at":
        key = lambda item: item.get("updated_at") or ""
    else:
        key = lambda item: item.get("stars") or 0

    return sorted(entries, key=key, reverse=reverse)


def _write_json(path: str, payload: Any) -> None:
    out_path = Path(path)
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(
        json.dumps(payload, indent=2, ensure_ascii=True) + "\n", encoding="utf-8"
    )


def _write_manifest(path: str, output_path: str, entries_count: int) -> None:
    output = Path(output_path)
    raw = output.read_bytes()
    checksum = hashlib.sha256(raw).hexdigest()
    payload = {
        "schema_version": 1,
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "entries": entries_count,
        "registry_path": output.name,
        "sha256": checksum,
    }
    _write_json(path, payload)


def _build_entry_for_repo(
    client: GitHubClient, config: Config, repo: str
) -> dict[str, Any]:
    repo_payload = client.fetch_repository(repo)
    if not repo_payload:
        raise RuntimeError("repository metadata not found")

    stars = repo_payload.get("stargazers_count")
    if isinstance(stars, int) and stars < config.min_stars:
        raise RuntimeError(f"below min_stars ({stars} < {config.min_stars})")

    if config.skip_archived and repo_payload.get("archived") is True:
        raise RuntimeError("repository archived")

    if config.skip_disabled and repo_payload.get("disabled") is True:
        raise RuntimeError("repository disabled")

    ref = repo_payload.get("default_branch")
    if not isinstance(ref, str) or not ref:
        ref = "HEAD"

    tree_items = client.fetch_repository_tree(repo, ref)
    colors = extract_colorschemes(tree_items)
    return build_entry(repo_payload, colors)


def run_once(config: Config) -> dict[str, int]:
    client = GitHubClient(
        request_delay_ms=config.request_delay_ms, retry_limit=config.retry_limit
    )
    store = StateStore(config.state_db_path)
    stats = {
        "discovered": 0,
        "fetched": 0,
        "cached": 0,
        "errors": 0,
        "written": 0,
    }

    try:
        discovered = discover_repositories(client, config)
        stats["discovered"] = len(discovered)

        entries_by_repo: dict[str, dict[str, Any]] = {}
        for payload in store.list_payloads():
            repo = payload.get("repo")
            if isinstance(repo, str) and repo:
                entries_by_repo[repo] = payload

        for repo, discovered_updated_at in discovered.items():
            if not store.should_refresh(
                repo, discovered_updated_at, config.stale_after_days
            ):
                cached = store.read_repo(repo)
                if cached and isinstance(cached.get("payload"), dict):
                    entries_by_repo[repo] = cached["payload"]
                    stats["cached"] += 1
                continue

            try:
                entry = _build_entry_for_repo(client, config, repo)
                updated_at = (
                    entry.get("updated_at")
                    if isinstance(entry.get("updated_at"), str)
                    else ""
                )
                store.upsert_repo(repo, updated_at, entry, parse_error=None)
                entries_by_repo[repo] = entry
                stats["fetched"] += 1
            except (GitHubRequestError, RuntimeError, ValueError) as error:
                store.upsert_repo(
                    repo,
                    discovered_updated_at or "",
                    {"repo": repo},
                    parse_error=str(error),
                )
                stats["errors"] += 1

        entries = list(entries_by_repo.values())
        overrides, excluded = load_overrides(config.overrides_path)
        merged = apply_overrides(entries, overrides, excluded)
        sorted_entries = _sort_entries(merged, config)

        _write_json(config.output_path, sorted_entries)
        _write_manifest(config.manifest_path, config.output_path, len(sorted_entries))
        stats["written"] = len(sorted_entries)
        return stats
    finally:
        store.close()


def run_loop(config: Config) -> None:
    while True:
        started = time.time()
        stats = run_once(config)
        took = int(time.time() - started)
        print(
            "run complete: "
            f"discovered={stats['discovered']} fetched={stats['fetched']} cached={stats['cached']} "
            f"errors={stats['errors']} written={stats['written']} duration={took}s"
        )
        time.sleep(config.scan_interval_seconds)

File: src/runner.ts
import { createHash } from "crypto";
import { mkdirSync, readFileSync, writeFileSync } from "fs";
import { dirname } from "path";
import type { ThemeEntry, GitHubRepoItem, GitHubTreeItem } from "./types.js";
import { validateThemeEntry } from "./types.js";
import type { Config } from "./config.js";
import { loadConfig, DEFAULT_CONFIG } from "./config.js";
import { GitHubClient, GitHubRequestError } from "./github-client.js";
import { StateStore } from "./state.js";
import { extractColorschemes, buildEntry } from "./parser.js";
import { loadOverrides, applyOverrides } from "./merge.js";
import { logger, setLogLevel } from "./logger.js";

export interface RunStats {
  discovered: number;
  scheduled: number;
  batches: number;
  fetched: number;
  cached: number;
  errors: number;
  written: number;
}

export function safeRepo(repo: string): string {
  return repo.trim().replace(/\.git$/, "").replace(/^\/+|\/+$/g, "");
}

async function discoverRepositories(
  client: GitHubClient,
  config: Config
): Promise<Map<string, string>> {
  const discovered = new Map<string, string>();

  for (const topic of config.topics) {
    logger.info(
      `discover topic=${topic} per_page=${config.per_page} max_pages_per_topic=${config.max_pages_per_topic}`
    );
    let page = 1;
    while (true) {
      const { items, hasNext } = await client.searchRepositories(
        topic,
        page,
        config.per_page
      );
      if (items.length === 0) break;

      for (const item of items) {
        const repo = safeRepo(item.full_name);
        if (repo && !discovered.has(repo)) {
          discovered.set(repo, item.updated_at);
        }
      }

      page++;
      if (config.max_pages_per_topic > 0 && page > config.max_pages_per_topic) {
        break;
      }
      if (!hasNext) break;
    }
  }

  for (const repo of config.include_repos) {
    const normalized = safeRepo(repo);
    if (normalized && !discovered.has(normalized)) {
      discovered.set(normalized, "");
    }
  }

  logger.info(
    `discover completed topics=${config.topics.length} repos=${discovered.size}`
  );
  return discovered;
}

export function selectRepositoriesForRun(
  discovered: Map<string, string>,
  config: Config
): Array<[string, string]> {
  const sorted = Array.from(discovered.entries()).sort((a, b) =>
    a[0].localeCompare(b[0])
  );
  if (config.max_repos_per_run > 0) {
    return sorted.slice(0, config.max_repos_per_run);
  }
  return sorted;
}

export function chunk<T>(items: T[], size: number): T[][] {
  const chunks: T[][] = [];
  for (let i = 0; i < items.length; i += size) {
    chunks.push(items.slice(i, i + size));
  }
  return chunks;
}

export function sortEntries(entries: ThemeEntry[], config: Config): ThemeEntry[] {
  const reverse = config.sort_order === "desc";

  return [...entries].sort((a, b) => {
    let cmp = 0;
    if (config.sort_by === "name") {
      cmp = (a.name || "").toLowerCase().localeCompare((b.name || "").toLowerCase());
    } else if (config.sort_by === "updated_at") {
      cmp = (a.updated_at || "").localeCompare(b.updated_at || "");
    } else {
      cmp = (a.stars || 0) - (b.stars || 0);
    }
    return reverse ? -cmp : cmp;
  });
}

export function writeJson(path: string, payload: unknown): void {
  const outPath = path;
  mkdirSync(dirname(outPath), { recursive: true });
  writeFileSync(outPath, JSON.stringify(payload, null, 2) + "\n", "utf-8");
}

export function writeManifest(
  manifestPath: string,
  outputPath: string,
  entriesCount: number
): void {
  const raw = readFileSync(outputPath);
  const checksum = createHash("sha256").update(raw).digest("hex");
  const payload = {
    count: entriesCount,
    generated_at: new Date().toISOString(),
    sha256: checksum,
  };
  writeJson(manifestPath, payload);
}

async function buildEntryForRepo(
  client: GitHubClient,
  config: Config,
  repo: string
): Promise<ThemeEntry> {
  const repoPayload = await client.fetchRepository(repo);
  if (!repoPayload) {
    throw new Error("repository metadata not found");
  }

  const stars = repoPayload.stargazers_count;
  if (typeof stars === "number" && stars < config.min_stars) {
    throw new Error(`below min_stars (${stars} < ${config.min_stars})`);
  }

  if (config.skip_archived && repoPayload.archived) {
    throw new Error("repository archived");
  }

  if (config.skip_disabled && repoPayload.disabled) {
    throw new Error("repository disabled");
  }

  const ref = repoPayload.default_branch || "HEAD";
  const treeItems = await client.fetchRepositoryTree(repo, ref);
  const colors = extractColorschemes(treeItems);
  return buildEntry(repoPayload, colors);
}

export async function runOnce(config: Config): Promise<RunStats> {
  const client = new GitHubClient({
    requestDelayMs: config.request_delay_ms,
    retryLimit: config.retry_limit,
  });
  const store = new StateStore(config.state_db_path);
  const stats: RunStats = {
    discovered: 0,
    scheduled: 0,
    batches: 0,
    fetched: 0,
    cached: 0,
    errors: 0,
    written: 0,
  };

  try {
    const discovered = await discoverRepositories(client, config);
    stats.discovered = discovered.size;
    const scheduled = selectRepositoriesForRun(discovered, config);
    stats.scheduled = scheduled.length;

    logger.info(
      `run plan discovered=${stats.discovered} scheduled=${stats.scheduled} batch_size=${config.batch_size} batch_pause_ms=${config.batch_pause_ms} request_delay_ms=${config.request_delay_ms}`
    );

    const entriesByRepo = new Map<string, ThemeEntry>();
    const persistedPayloads = store.listPayloads();
    for (const payload of persistedPayloads) {
      const repo = payload.repo;
      if (repo) {
        entriesByRepo.set(repo, payload);
      }
    }
    logger.debug(`loaded payloads from state count=${persistedPayloads.length}`);

    const batchGroups = chunk(scheduled, config.batch_size);
    const totalBatches = batchGroups.length;

    for (let batchIndex = 0; batchIndex < batchGroups.length; batchIndex++) {
      const batch = batchGroups[batchIndex];
      if (!batch) continue;
      stats.batches++;
      logger.info(
        `processing batch=${batchIndex + 1}/${totalBatches} size=${batch.length}`
      );

      for (const [repo, discoveredUpdatedAt] of batch) {
        if (!store.shouldRefresh(repo, discoveredUpdatedAt, config.stale_after_days)) {
          const cached = store.readRepo(repo);
          if (cached?.payload && "repo" in cached.payload) {
            entriesByRepo.set(repo, cached.payload as ThemeEntry);
            stats.cached++;
          }
          continue;
        }

        try {
          const entry = await buildEntryForRepo(client, config, repo);
          const updatedAt = entry.updated_at || "";
          store.upsertRepo(repo, updatedAt, entry, null);
          entriesByRepo.set(repo, entry);
          stats.fetched++;
        } catch (error) {
          const errorMessage =
            error instanceof Error ? error.message : String(error);
          store.upsertRepo(repo, discoveredUpdatedAt || "", { repo }, errorMessage);
          stats.errors++;
          logger.warn(`repo processing failed repo=${repo} error=${errorMessage}`);
        }
      }

      const entries = Array.from(entriesByRepo.values());
      const { overrides, excluded } = loadOverrides(config.overrides_path);
      const merged = applyOverrides(entries, overrides, excluded);
      const sortedEntries = sortEntries(merged, config);

      const validEntries: ThemeEntry[] = [];
      for (const entry of sortedEntries) {
        const validated = validateThemeEntry(entry);
        if (validated) {
          validEntries.push(validated);
        }
      }

      writeJson(config.output_path, validEntries);
      writeManifest(config.manifest_path, config.output_path, validEntries.length);
      stats.written = validEntries.length;
      logger.debug(
        `batch checkpoint written batch=${batchIndex + 1}/${totalBatches} entries=${validEntries.length}`
      );

      if (config.batch_pause_ms > 0 && batchIndex < totalBatches - 1) {
        const pauseSeconds = config.batch_pause_ms / 1000;
        logger.debug(`batch pause sleep=${pauseSeconds}s`);
        await new Promise((resolve) => setTimeout(resolve, config.batch_pause_ms));
      }
    }

    logger.info(
      `run complete discovered=${stats.discovered} scheduled=${stats.scheduled} batches=${stats.batches} fetched=${stats.fetched} cached=${stats.cached} errors=${stats.errors} written=${stats.written}`
    );

    return stats;
  } finally {
    store.close();
  }
}

export async function runLoop(config: Config): Promise<void> {
  while (true) {
    const started = Date.now();
    logger.info("loop iteration started");
    const stats = await runOnce(config);
    const took = Math.floor((Date.now() - started) / 1000);
    logger.info(`loop iteration finished duration=${took}s stats=${JSON.stringify(stats)}`);
    await new Promise((resolve) =>
      setTimeout(resolve, config.scan_interval_seconds * 1000)
    );
  }
}

export { Config, loadConfig, DEFAULT_CONFIG, setLogLevel };
